### ISSUE TITLE
Is `NotParameterizedAttribute` working as expected?

### ISSUE THREAD

**Issue by azygis** [dotnet/efcore#24742]
I'm not sure whether it's a bug report or just misunderstanding, but based on the attribute name and/or xmldocs, it should not be working correctly.

Consider the following setup:

```
// ...
protected override void OnModelCreating(ModelBuilder modelBuilder)
{
    modelBuilder.HasDbFunction(() => FromPostgis(default, default))
        .HasTranslation(CustomSqlTranslatingExpressionVisitor.TranslateFromPostgis);
// ...
}

public virtual byte[] ToPostgis([NotParameterized] GeoFormat geoFormat, [NotParameterized] string data) =>
    throw new NotSupportedException(_errorMessage);
```

We have such query:
```
var result = await _context.WithoutEntity
    .Select(_ => new PlainGeometry
    {
        Format = query.Format,
        Data = _context.FromPostgis(query.Format,
            _context.ST_Difference(
                _context.ST_MakeValid(query.Geometry),
                _context.ST_MakeValid(query.CutterGeometry)))
    })
    .FirstOrDefaultAsync();
```

In the translation method, the very beginning:
```
public static SqlExpression TranslateFromPostgis(IReadOnlyCollection<SqlExpression> expressions)
{
    var geoFormat = (GeoFormat) ((SqlConstantExpression) expressions.First()).Value;
    // ...
}
```

The above throws `InvalidCastException` because it's not possible to cast from `SqlParameterExpression` to `SqlConstantExpression`. Since we need to have the value for our translation, we are forced to change most of our queries to never include the same variable/property twice in the same query (which is good enough for workaround, but we'd rather not do that), e.g.:

```
var result = await _context.WithoutEntity
    .Select(_ => new PlainGeometry
    {
        Data = _context.FromPostgis(query.Format,
            _context.ST_Difference(
                _context.ST_MakeValid(query.Geometry),
                _context.ST_MakeValid(query.CutterGeometry)))
    })
    .FirstOrDefaultAsync();
result.Format = query.Format;
```

As soon as the same variable or property is used multiple times in the same query, expression gets parameterized.

So my question is... Does the `NotParameterizedAttribute` actually work as expected? Should it not **force** function argument to not be `SqlParameterExpression`, regardless of whether the same value is used in multiple places?

**Comment by smitpatel** [comment]
You are registering and using function `FromPostgis` in query but your `NotParameterizedAttribute` is used on function named `ToPostgis`.

**Comment by azygis** [comment]
Sorry, copied the wrong method. `FromPostgis` has exactly the same signature. Frankly, both methods behave the same as soon as there's any variable/property reuse in any place in the expression.

**Comment by smitpatel** [comment]
Please provide a full runnable repro code.

**Comment by azygis** [comment]
Before I do that - am I correct when thinking that the attribute should actually force parameters being `SqlConstantExpression` instead of `SqlParameterExpression`? Time is of the essence, so if I fell in the case where it can't simply be constant, I would not make any trouble neither for you guys nor myself. If indeed it should be constants with the attribute - I will create a repro.

**Comment by smitpatel** [comment]
The attribute is supposed to be used when a closure dependent variable is used as a parameter and it shouldn't be generated as a query parameter. There is no guarantee of it being SqlConstantExpression always. Without looking at repro code, we cannot say that it should be SqlConstantExpression always.

**Comment by azygis** [comment]
Okay, I'll create the repro in the following days.

**Comment by azygis** [comment]
[Here's the repro code](https://github.com/dotnet/efcore/files/6374970/NotParameterizedAttributeIssue.zip).

**Comment by azygis** [comment]
Apparently, order of assignments matters for some reason (not gonna lie, I'm not very (or at all) good at how expressions work):

```
private static void Ok3(GetPostQuery query)
{
    using (var db = new DatabaseContext())
    {
        var post = db.Posts
            .Select(x =>
                new
                {
                    Data = db.FromPostgis(query.Format, x.Content),
                    Format = query.Format
                })
            .ToQueryString();
    }
}
```
This way it's fine, but if format / data assignments have places switched, it's broken. This makes me believe there is indeed a bug somewhere.

**Comment by smitpatel** [comment]
Which version of EF Core are you using?
Ordering changed things because query.Format is already parameterized and while seeing it again, we just reused parameter.

**Comment by azygis** [comment]
Version is 5.0.2 in our repo. Not sure which version is in zip file, on mobile now and my colleague created it today. Should be something similar.

As mentioned, we found the workaround in our case, but that seems rather cumbersome and requires wiki-like page explaining other devs how to not make the same mistake we did.

Using the attribute could overrule the parameter if it's in a later stage of expression, maybe?

**Comment by smitpatel** [comment]
We may have fixed this in 6.0-preview3

**Comment by azygis** [comment]
Indeed, in preview the order doesn't matter and works as expected. We'll wait for a stable release then, thank you!

**Comment by roji** [comment]
@smitpatel do we want to use this to track the previous bug (or deup it with another issue)?
